connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamingQueryCache.scala

```diff
@@ -0,0 +1,203 @@
+ *   - Keep the session and session mapping at connect server alive as long as a streaming query
+ *     is active. Even if the client side has disconnected.
+ *     - This matches how streaming queries behave in Spark. The queries continue to run if
+ *       notebook or job session is lost.
+ *   - Once a query is stopped, the reference and mappings are maintained for 1 hour and will be
+ *     accessible from the client. This allows time for client to fetch status.
+ *     - During this time if the query is restarted (i.e. has a new run id), the reference to
+ *       previous run is dropped. As a result logical query has only the most recent query
+ *       reference cached. This policy can be revisited to cache multiple runs for a query.
+ *   - Note that these semantics are evolving and might change before being finalized in Connect.
+ *   - Future improvements:
+ *     - Provide an API for a users to access session even after they lose session id.
+ *       - Once a user is properly authenticated, the API could return list of sessions that are
+ *         still alive in the connect server for that user.
+ */
+private[connect] class SparkConnectStreamingQueryCache(
+    val sessionKeepAliveFn: (String, String) => Unit, // (userId, sessionId) => Unit.
+    val clock: Clock = new SystemClock(),
+    private val stoppedQueryCachePeriod: Duration = 1.hour, // Configurable for testing.
+    private val sessionPollingPeriod: Duration = 1.minute // Configurable for testing.
+) extends Logging {
+
+  import SparkConnectStreamingQueryCache._
+
+  def registerNewStreamingQuery(sessionHolder: SessionHolder, query: StreamingQuery): Unit = {
+    queryCacheLock.synchronized {
+      val value = QueryCacheValue(
+        userId = sessionHolder.userId,
+        sessionId = sessionHolder.sessionId,
+        session = sessionHolder.session,
+        query = query,
+        expiresAtMs = None)
+
+      queryCache.put(QueryCacheKey(query.id.toString), value) match {
+        case Some(existing) => // Query is being replaced. Can happen when a query is restarted.
+          log.info(
+            s"Replacing existing query in the cache. Query Id: ${query.id}." +
+              s"Existing value $existing, new value $value.")
+        case None =>
+          log.info(s"Adding new query to the cache. Query Id ${query.id}, value $value.")
+      }
+
+      schedulePeriodicChecks() // Starts the scheduler thread if it hasn't started.
+    }
+  }
+
+  /**
+   * Returns [[StreamingQuery]] if it is cached and session matches the cached query. It ensures
+   * the the session associated with it matches the session passed into the call.
+   */
+  def findCachedQuery(queryId: String, session: SparkSession): Option[StreamingQuery] = {
+    queryCacheLock.synchronized {
+      cacheQueryValue(queryId).flatMap { v =>
+        if (v.session == session) Some(v.query)
+        else None // This should be rare. Likely the query is restarted on a different session.
```

Comment:
Just trying to understand how this can happen. If it's restarted on a different session, a new value will replace existing one with same queryId, then it can still be found from the cache, right? Or are u referring to possible race condition here?